### PR TITLE
family: update nrf52 approtect handling

### PR DESCRIPTION
**Pull request by maxd-nordic**
This patch handles the updated APPROTECT mechanism that is implemented in newer revisions of the nRF52 devices. Details on that are described here: https://devzone.nordicsemi.com/nordic/nordic-blog/b/blog/posts/working-with-the-nrf52-series-improved-approtect This is compatible with both old and new revisions.

### CODE REVIEW

**Review by flit — APPROVED** [review]

**Comment by flit** [comment]
Looks like all the functional tests on the micro:bit are failing. You can see the [Basic Test results here](https://dev.azure.com/pyocd/pyocd/_build/results?buildId=415&view=ms.vss-test-web.build-test-results-tab&runId=2094&resultId=100042&paneView=debug). 

The important part of the log:
```
reset and halt
Exception cannot read register pc because core #0 is not halted when testing board 9904360258994e450053900b00000041000000009796990b
Traceback (most recent call last):
  File "C:\Users\ci\azpagent\_work\1\s\test\test_util.py", line 270, in run
    self.test_function(board.unique_id)
  File "C:\Users\ci\azpagent\_work\1\s\test\basic_test.py", line 58, in run_basic_test
    return basic_test(board_id, None)
  File "C:\Users\ci\azpagent\_work\1\s\test\basic_test.py", line 148, in basic_test
    currentPC = target.read_core_register('pc')
  File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\core\soc_target.py", line 220, in read_core_register
    return self.selected_core_or_raise.read_core_register(id)
  File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 995, in read_core_register
    regValue = self.read_core_register_raw(reg_info.index)
  File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 1012, in read_core_register_raw
    vals = self.read_core_registers_raw([reg])
  File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 1033, in read_core_registers_raw
    return self._base_read_core_registers_raw(reg_list)
  File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 1047, in _base_read_core_registers_raw
    raise exceptions.CoreRegisterAccessError(
pyocd.core.exceptions.CoreRegisterAccessError: cannot read register pc because core #0 is not halted
```

**Comment by flit** [comment]
Just noticed, would you mind changing the target branch for this PR to `develop`? Since it's more than just a simple bug fix. Thanks.

**Comment by flit** [comment]
Forgot something in the review: would you mind adding Nordic's copyright to the file's header comment? Thanks!

File: pyocd/target/family/target_nRF52.py
Comment (flit) [inline pyocd/target/family/target_nRF52.py:1]: Please add Nordic's copyright to the file header comment. Thanks!

Comment (maxd-nordic) [reply pyocd/target/family/target_nRF52.py:1]: done

**Comment by maxd-nordic** [comment]
> Looks like all the functional tests on the micro:bit are failing. You can see the [Basic Test results here](https://dev.azure.com/pyocd/pyocd/_build/results?buildId=415&view=ms.vss-test-web.build-test-results-tab&runId=2094&resultId=100042&paneView=debug).
> 
> The important part of the log:
> 
> ```
> reset and halt
> Exception cannot read register pc because core #0 is not halted when testing board 9904360258994e450053900b00000041000000009796990b
> Traceback (most recent call last):
>   File "C:\Users\ci\azpagent\_work\1\s\test\test_util.py", line 270, in run
>     self.test_function(board.unique_id)
>   File "C:\Users\ci\azpagent\_work\1\s\test\basic_test.py", line 58, in run_basic_test
>     return basic_test(board_id, None)
>   File "C:\Users\ci\azpagent\_work\1\s\test\basic_test.py", line 148, in basic_test
>     currentPC = target.read_core_register('pc')
>   File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\core\soc_target.py", line 220, in read_core_register
>     return self.selected_core_or_raise.read_core_register(id)
>   File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 995, in read_core_register
>     regValue = self.read_core_register_raw(reg_info.index)
>   File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 1012, in read_core_register_raw
>     vals = self.read_core_registers_raw([reg])
>   File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 1033, in read_core_registers_raw
>     return self._base_read_core_registers_raw(reg_list)
>   File "C:\Users\ci\azpagent\_work\1\s\.venv\lib\site-packages\pyocd\coresight\cortex_m.py", line 1047, in _base_read_core_registers_raw
>     raise exceptions.CoreRegisterAccessError(
> pyocd.core.exceptions.CoreRegisterAccessError: cannot read register pc because core #0 is not halted
> ```

Ah, sorry. I thought I could skip the HW revision check because it worked on an old nrf52840. Added that now.

**Review by flit — APPROVED** [review]

**Pull request merged**